qabelbox/src/main/java/de/qabel/qabelbox/services/LocalQabelService.java

```diff
@@ -522,13 +531,39 @@ public void onCreate() {
         Log.i(TAG, "LocalQabelService created");
         dropHTTP = new DropHTTP();
         initSharedPreferences();
-        initAndroidPersistence();
+        initRepositories();
+        try {
+            migratePersistence();
+        } catch (EntityNotFoundExcepion | PersistenceException e) {
+            Log.e(TAG, "Migration of identities and contatcs failed", e);
+        }
         pendingUploads = new HashMap<>();
         documentIdParser = new DocumentIdParser();
-        cachedFinishedUploads = Collections.synchronizedMap(new HashMap<String, Map<String, BoxFile>>());
+        cachedFinishedUploads = Collections.synchronizedMap(new HashMap<>());
         uploadingQueue = new LinkedBlockingDeque<>();
-        self = this;
+
     }
+
+    private void migratePersistence() throws EntityNotFoundExcepion, PersistenceException {
+        initAndroidPersistence();
+        PersistenceMigration.migrate(persistence, identityRepository, contactRepository);
+    }
+
+    public AndroidPersistence getPersistence() {
+        return persistence;
+    }
+
+    public void initRepositories() {
+        RepositoryFactory repositoryFactory = new RepositoryFactory(getApplicationContext());
+        try {
+            AndroidClientDatabase androidClientDatabase = repositoryFactory.getAndroidClientDatabase();
+            identityRepository = repositoryFactory.getIdentityRepository(androidClientDatabase);
+            contactRepository = repositoryFactory.getContactRepository(androidClientDatabase);
+        } catch (SQLException e) {
```

Comment:
you're polluting the application with SQL specific exceptions. I would prefer finding a way to avoid that (like not calling getAndroidClientDatabase() externally) so that this details gets abstracted away by the repository layer... especially because you're just rethrowing RuntimeExceptions most of the time (so the `SQL` part really doesn't matter)